packages/@aws-cdk/aws-s3/lib/bucket.ts

```diff
@@ -776,6 +780,31 @@ export class Bucket extends BucketBase {
     return this.onEvent(EventType.ObjectRemoved, dest, ...filters);
   }
 
+  private validateBucketName(bucketName: string) {
+    const errors: string[] = [];
+
+    // Rules codified from https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html
+    if (bucketName.length < 3 || bucketName.length > 63) {
+      errors.push('Bucket name must be at least 3 and no more than 63 characters');
+    }
+    if (!/^[a-z0-9.-]+$/.test(bucketName)) {
+      errors.push('Bucket name must only contain lowercase characters and the symbols, period (.) and dash (-)');
+    }
+    if (!/[a-z0-9]/.test(bucketName.charAt(0)) || !/[a-z0-9]/.test(bucketName.charAt(bucketName.length - 1))) {
+      errors.push('Bucket name must start and end with a lowercase character or number');
+    }
+    if (/\.-|-\.|\.\./.test(bucketName)) {
+      errors.push('Bucket name must not have dash next to period, or period next to dash, or consecutive periods');
+    }
+    if (/\d+\.\d+\.\d+\.\d+/.test(bucketName)) {
```

Comment:
Is this how S3 validates? I feel like this might be prohibiting certain options that S3 would permit...

Reply:
Not sure if this is how S3 does it. I've reversed engineered their validations based on their [rules here](https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html) and manually trying out combinations.

Some of these are pretty straight forward, like length and character set, but the ones around IP and symbol combinations get more tricky. 
I've picked combinations and regex that I should be more liberal than S3 (in that, S3 might have additional/stricter checks I don't have here), but happy to drop the more tricky ones if you think they might break customers.

What do you think?